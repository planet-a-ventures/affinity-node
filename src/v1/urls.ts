/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#get-a-specific-list) for more info.
 */
export const listsUrl = (list_id?: number) =>
    list_id ? `/lists/${encodeURIComponent(list_id)}` : '/lists'

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#get-a-specific-list-entry) for more info.
 */
export const listEntriesUrl = (list_id: number, list_entry_id?: number) =>
    listsUrl(list_id) + (
        list_entry_id
            ? `/list-entries/${encodeURIComponent(list_entry_id)}`
            : '/list-entries'
    )

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#rate-limit) for more info.
 */
export const rateLimitUrl = () => '/rate-limit'

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#whoami) for more info.
 */
export const whoAmIUrl = () => '/auth/whoami'

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#fields) for more info.
 */
export const fieldsUrl = (field_id?: number) =>
    field_id ? `/fields/${encodeURIComponent(field_id)}` : '/fields'

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#field-values) for more info.
 */
export const fieldValuesUrl = (field_value_id?: number) => {
    return field_value_id
        ? `/field-values/${encodeURIComponent(field_value_id)}`
        : '/field-values'
}

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#get-field-values-changes) for more info.
 */
export const fieldValueChangesUrl = () => '/field-value-changes'

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#organizations) for more info.
 */
export const organizationsUrl = (organization_id?: number | 'fields') => {
    return organization_id
        ? `/organizations/${encodeURIComponent(organization_id)}`
        : '/organizations'
}

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#get-global-organizations-fields) for more info.
 */
export const organizationFieldsUrl = () => organizationsUrl('fields')

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#persons) for more info.
 */
export const personsUrl = (person_id?: number | 'fields') => {
    return person_id ? `/persons/${encodeURIComponent(person_id)}` : '/persons'
}

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#get-global-person-fields) for more info.
 */
export const personFieldsUrl = () => personsUrl('fields')

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#entity-files) for more info.
 */
export const entityFilesUrl = (
    entity_file_id?: number,
    is_download: boolean = false,
) => {
    return entity_file_id
        ? `/entity-files` + (is_download ? '/download' : '') +
            `/${encodeURIComponent(entity_file_id)}`
        : '/entity-files'
}
