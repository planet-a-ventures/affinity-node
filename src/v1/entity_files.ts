import { assert } from '@std/assert'
import type { AxiosInstance } from 'axios'
import { fileFrom } from 'fetch-blob/from.js'
import type { Readable } from 'node:stream'
import { defaultTransformers } from './axios_default_transformers.ts'
import { createSearchIteratorFn } from './create_search_iterator_fn.ts'
import { EntityRequestFilter } from './field_value_changes.ts'
import type { PagedRequest } from './paged_request.ts'
import type { PagedResponse } from './paged_response.ts'
import type { DateTime, Replace, RequireOnlyOne } from './types.ts'
import { entityFilesUrl } from './urls.ts'
export type { DateTime } from './types.ts'

export type EntityFileRaw = {
    /** The unique identifier of the entity file object. */
    id: number
    /** The name of the file. */
    name: string
    /** The size of the file in bytes. */
    size: number
    /** The unique identifier of the person corresponding to the entity file. */
    person_id: number | null
    /** The unique identifier of the organization corresponding to the entity file. */
    organization_id: number | null
    /** The unique identifier of the opportunity corresponding to the entity file. */
    opportunity_id: number | null
    /** The unique identifier of the user who created the entity file. */
    uploader_id: number
    /** The time when the entity file was created. */
    created_at: DateTime
}

export type EntityFile = Replace<EntityFileRaw, {
    created_at: Date
}>

/**
 * Represents the request parameters for retrieving entity files.
 */
export type AllEntityFileRequest =
    & {
        /**
         * A unique ID that represents a Person whose associated files should be retrieved.
         */
        person_id?: number
        /**
         * A unique ID that represents an Organization whose associated files should be retrieved.
         */
        organization_id?: number
        /**
         * A unique ID that represents an Opportunity whose associated files should be retrieved.
         */
        opportunity_id?: number
    }
    & PagedRequest

export type PagedEntityFileResponseRaw =
    & {
        entity_files: EntityFileRaw[]
    }
    & PagedResponse

export type PagedEntityFileResponse = Replace<PagedEntityFileResponseRaw, {
    entity_files: EntityFile[]
}>

export type SupportedFileType = File | string

export type UploadEntityFileRequest =
    & {
        files: SupportedFileType[]
    }
    & RequireOnlyOne<EntityRequestFilter, keyof EntityRequestFilter>

/**
 * Entity files are files uploaded to a relevant entity.
 * Possible files, for example, would be a pitch deck for an opportunity or a physical mail correspondence for a person.
 */
export class EntityFiles {
    /** @hidden */
    constructor(private readonly axios: AxiosInstance) {
    }

    private static transformEntityFile(file: EntityFileRaw): EntityFile {
        return {
            ...file,
            created_at: new Date(file.created_at),
        }
    }

    /**
     * Returns all entity files within your organization.
     */
    async all(params?: AllEntityFileRequest): Promise<PagedEntityFileResponse> {
        const response = await this.axios.get<PagedEntityFileResponse>(
            entityFilesUrl(),
            {
                params,
                transformResponse: [
                    ...defaultTransformers(),
                    (json: PagedEntityFileResponseRaw) => {
                        return {
                            ...json,
                            entity_files: json.entity_files.map(
                                EntityFiles.transformEntityFile,
                            ),
                        }
                    },
                ],
            },
        )
        return response.data
    }

    /**
     * Returns an async iterator that yields all entity files matching the given request
     * Each yielded array contains up to the number specified in {@link AllEntityFileRequest.page_size} of entity files.
     * Use this method if you want to process the entity files in a streaming fashion.
     *
     * *Please note:* the yielded entity files array may be empty on the last page.
     *
     * @example
     * ```typescript
     * let page = 0
     * for await (const entries of affinity.entityFiles.pagedIterator({
     *     person_id: 123,
     *     page_size: 10
     * })) {
     *     console.log(`Page ${++page} of entries:`, entries)
     * }
     * ```
     */
    pagedIterator = createSearchIteratorFn(
        this.all.bind(this),
        'entity_files',
    )

    /**
     * Fetches an entity with a specified `entity_file_id`.
     */
    async get(entity_file_id: EntityFile['id']): Promise<EntityFile> {
        const response = await this.axios.get<EntityFile>(
            entityFilesUrl(entity_file_id),
            {
                transformResponse: [
                    ...defaultTransformers(),
                    EntityFiles.transformEntityFile,
                ],
            },
        )
        return response.data
    }

    /**
     * Downloads the entity file with the specified `entity_file_id`.
     *
     * @example
     * ```typescript
     * import { promises as fsPromises } from 'node:fs';
     * const fileResponseStream = affinity.entityFiles.download(123);
     * await fsPromises.writeFile(filePath, fileResponseStream);
     * ```
     */
    async download(entity_file_id: EntityFile['id']): Promise<Readable> {
        const response = await this.axios.get(
            entityFilesUrl(entity_file_id, true),
            {
                responseType: 'stream',
                // The download location of entity files is provided via a redirect from Affinity
                maxRedirects: 5,
            },
        )
        return response.data
    }

    /**
     * Uploads files attached to the entity with the given id.
     *
     * The file will display on the entity's profile, provided that the entity is not a person internal to the user's organization.
     *
     * @example
     * ```typescript
     * const file = fs.createReadStream('example.pdf');
     * const entityFile = await affinity.entityFiles.upload({
     *     files: [file],
     *     person_id: 123,
     * });
     * ```
     */
    async upload(params: UploadEntityFileRequest): Promise<boolean> {
        const formData = new FormData()
        const { files } = params
        assert(files.length, 'At least one file must be provided')

        if (params.person_id) {
            formData.append('person_id', params.person_id.toString())
        } else if (params.organization_id) {
            formData.append(
                'organization_id',
                params.organization_id.toString(),
            )
        } else if (params.opportunity_id) {
            formData.append('opportunity_id', params.opportunity_id.toString())
        } else {
            throw new Error(
                'One of person_id, organization_id or opportunity_id must be provided',
            )
        }

        await Promise.all(
            files.map((file) => appendToFormData(formData, file)),
        )
        const response = await this.axios.post<{ success: boolean }>(
            entityFilesUrl(),
            formData,
            {
                headers: {
                    'Content-Type': 'multipart/form-data',
                },
            },
        )
        return response.data.success === true
    }
}

function isFile(file: unknown): file is File {
    return file instanceof File
}

async function appendToFormData(
    formData: FormData,
    file: SupportedFileType,
) {
    if (typeof file === 'string') {
        const f = await fileFrom(file)
        formData.append('files[]', f)
    } else if (isFile(file)) {
        formData.append('files[]', file)
        return null
    } else {
        throw new Error('Unsupported file type')
    }
}
