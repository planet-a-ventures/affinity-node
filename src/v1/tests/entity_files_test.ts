import { assert, assertEquals } from '@std/assert'
import * as path from '@std/path'
import { afterEach, beforeEach, describe, it } from '@std/testing/bdd'
import { assertSnapshot } from '@std/testing/snapshot'
import axios from 'axios'
import MockAdapter from 'axios-mock-adapter'
import { Buffer } from 'jsr:@std/io/buffer'
import { Readable } from 'node:stream'
import { buffer } from 'node:stream/consumers'
import { AllEntityFileRequest } from '../entity_files.ts'
import { Affinity } from '../index.ts'
import { entityFilesUrl } from '../urls.ts'
import { apiKey, isLiveRun } from './env.ts'
import { getRawFixture, readFixtureFile } from './get_raw_fixture.ts'
import { mockPagingFromAllEndpoint } from './mock_paging_from_all_endpoint.ts'

const __dirname = path.dirname(path.fromFileUrl(import.meta.url))

const multipartFormDataHeaderMatcher = {
    headers: {
        asymmetricMatch: (headers: Record<string, string>) => {
            assertEquals(headers['Content-Type'], 'multipart/form-data')
            return true
        },
    },
}

const createSnapshotBodyMatcher = (t: Deno.TestContext) => ({
    asymmetricMatch: async (reqBody: FormData) => {
        const data: Record<string, unknown> = Object.fromEntries(
            reqBody.entries(),
        )
        if (reqBody.has('files[]')) {
            // normal serialization overwrites duplicated keys, so we need to handle this case
            data['files[]'] = reqBody.getAll('files[]')
        }
        await assertSnapshot(t, data)
        return true
    },
})

describe('entityFiles', () => {
    let mock: MockAdapter
    let affinity: Affinity

    beforeEach(() => {
        if (!isLiveRun()) {
            mock = new MockAdapter(axios, { onNoMatch: 'throwException' })
        }
        affinity = new Affinity(apiKey() || 'api_key')
    })
    afterEach(() => {
        mock?.reset()
    })

    it('can list all files', async (t) => {
        mock?.onGet(entityFilesUrl()).reply(
            200,
            await getRawFixture('entity_files/all.raw.response.json'),
        )
        const res = await affinity.entityFiles.all({
            person_id: 170614434,
        })
        await assertSnapshot(t, res)
    })

    it('can get a single file', async (t) => {
        mock?.onGet(entityFilesUrl(131)).reply(
            200,
            await getRawFixture('entity_files/get.raw.response.json'),
        )
        const res = await affinity.entityFiles.get(131)
        await assertSnapshot(t, res)
    })

    it('can upload files from path', async (t) => {
        mock
            ?.onPost(
                entityFilesUrl(),
                createSnapshotBodyMatcher(t),
                multipartFormDataHeaderMatcher,
            )
            .reply(
                200,
                { success: true },
            )
        const localPath = path.join(
            __dirname,
            'fixtures',
            'entity_files',
            'test.pdf',
        )
        const res = await affinity.entityFiles.upload({
            person_id: 170614434,
            files: [
                localPath,
            ],
        })
        assert(res)
    })

    it('can upload a file', async (t) => {
        mock
            ?.onPost(
                entityFilesUrl(),
                createSnapshotBodyMatcher(t),
                multipartFormDataHeaderMatcher,
            )
            .reply(
                200,
                { success: true },
            )
        const res = await affinity.entityFiles.upload({
            person_id: 170614434,
            files: [
                new File(
                    [
                        await readFixtureFile('./entity_files/test.pdf'),
                    ],
                    'test.pdf',
                ),
            ],
        })
        assert(res)
    })

    it('can upload multiple files', async (t) => {
        mock
            ?.onPost(
                entityFilesUrl(),
                createSnapshotBodyMatcher(t),
                multipartFormDataHeaderMatcher,
            )
            .reply(
                200,
                { success: true },
            )
        const res = await affinity.entityFiles.upload({
            person_id: 170614434,
            files: [
                new File(
                    [await readFixtureFile('./entity_files/test.pdf')],
                    'test1.pdf',
                ),
                new File(
                    [await readFixtureFile('./entity_files/test.pdf')],
                    'test2.pdf',
                ),
            ],
        })
        assert(res)
    })

    it('can download a file', async (t) => {
        const pdfContents = await readFixtureFile('./entity_files/test.pdf')

        mock?.onGet(entityFilesUrl(6534776, true)).reply(
            200,
            readableFromArray(pdfContents),
        )

        const stream = await affinity.entityFiles.download(6534776)
        const buf: ArrayBuffer = await buffer(stream)

        const expected = new Buffer()
        expected.read(pdfContents)
        assertEquals(new Buffer(buf), expected)
    })

    it('iterates over all entity files', async (t) => {
        const params: AllEntityFileRequest = {
            person_id: 142,
            page_size: 1,
        }

        await mockPagingFromAllEndpoint(
            './fixtures/entity_files/all.raw.response.json',
            params,
            'entity_files',
            entityFilesUrl,
            mock,
        )

        let page = 0
        for await (
            const entries of affinity.entityFiles.pagedIterator(params)
        ) {
            await assertSnapshot(t, entries, {
                name: `page ${++page} of entity files`,
            })
        }
    })
})

function readableFromArray(arr: Uint8Array) {
    return new Readable({
        read(size: number) {
            this.push(arr)
            this.push(null)
        },
    })
}
