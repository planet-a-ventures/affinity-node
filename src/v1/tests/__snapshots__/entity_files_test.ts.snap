export const snapshot = {};

snapshot[`entityFiles > can list all files 1`] = `
{
  entity_files: [
    {
      created_at: 2011-01-25T17:59:35.288Z,
      id: 43212,
      name: "JohnDoeFriends.csv",
      opportunity_id: null,
      organization_id: null,
      person_id: 142,
      size: 993,
      uploader_id: 10,
    },
    {
      created_at: 2019-01-13T20:52:51.539Z,
      id: 131,
      name: "Import.csv",
      opportunity_id: null,
      organization_id: null,
      person_id: 38654,
      size: 227224,
      uploader_id: 101,
    },
  ],
  next_page_token: null,
}
`;

snapshot[`entityFiles > can get a single file 1`] = `
{
  created_at: 2019-01-13T20:52:51.539Z,
  id: 131,
  name: "Import.csv",
  opportunity_id: null,
  organization_id: null,
  person_id: 38654,
  size: 227224,
  uploader_id: 101,
}
`;

snapshot[`entityFiles > can upload files from path 1`] = `
{
  "files[]": [
    MyFile {
      name: "test.pdf",
      size: 16374,
      type: "",
    },
  ],
  person_id: "170614434",
}
`;

snapshot[`entityFiles > can upload a file 1`] = `
{
  "files[]": [
    File {
      name: "test.pdf",
      size: 16374,
      type: "",
    },
  ],
  person_id: "170614434",
}
`;

snapshot[`entityFiles > can upload multiple files 1`] = `
{
  "files[]": [
    File {
      name: "test1.pdf",
      size: 16374,
      type: "",
    },
    File {
      name: "test2.pdf",
      size: 16374,
      type: "",
    },
  ],
  person_id: "170614434",
}
`;

snapshot[`page 1 of entity files 1`] = `
[
  {
    created_at: 2011-01-25T17:59:35.288Z,
    id: 43212,
    name: "JohnDoeFriends.csv",
    opportunity_id: null,
    organization_id: null,
    person_id: 142,
    size: 993,
    uploader_id: 10,
  },
]
`;

snapshot[`page 2 of entity files 1`] = `
[
  {
    created_at: 2019-01-13T20:52:51.539Z,
    id: 131,
    name: "Import.csv",
    opportunity_id: null,
    organization_id: null,
    person_id: 142,
    size: 227224,
    uploader_id: 101,
  },
]
`;

snapshot[`page 3 of entity files 1`] = `[]`;
